File: implementations/python/benchmark/scheduled_gate.py
"""Revision 4 full-resolution scheduled-control acceptance gate."""

import json
from pathlib import Path
from typing import cast

import numpy as np

from foilbench_py.core.geometry import NacaFoil
from foilbench_py.core.scenario import find_repo_root, load_scenario
from foilbench_py.solvers.factory import create_solver


def main() -> None:
    root = find_repo_root(Path(__file__))
    fixture = cast(
        dict[str, object],
        json.loads(
            (root / "spec/conformance/fullsize-acceptance.json").read_text(
                encoding="utf-8"
            )
        ),
    )
    gate = cast(dict[str, object], fixture["scheduled_checkpoints"])
    scenario = load_scenario(root / str(gate["scenario"]))
    resolution = tuple(cast(list[int], gate["resolution"]))
    if scenario.domain.resolution != resolution:
        raise ValueError("scheduled fixture resolution disagrees with its scenario")
    solver = create_solver(str(gate["solver"]))
    solver.initialize(scenario, NacaFoil(scenario.foil), scenario.seed)
    checkpoints = [float(value) for value in cast(list[float], gate["times"])]
    checkpoint_steps = {round(value / scenario.output_dt): value for value in checkpoints}
    for step in range(1, max(checkpoint_steps) + 1):
        time = step * scenario.output_dt
        solver.advance(scenario.control_at(time), scenario.output_dt)
        if step in checkpoint_steps:
            state = solver.export_state()
            if not np.isfinite(state.velocity).all():
                raise RuntimeError(f"non-finite scheduled state at t={time}")
            print(f"passed scheduled checkpoint t={time:.2f}")


if __name__ == "__main__":
    main()

File: implementations/julia/benchmark/scheduled_gate.jl
#!/usr/bin/env julia

using FoilBenchJulia
using JSON3

function main()
    root = find_repository_root(@__DIR__)
    fixture = JSON3.read(read(joinpath(root, "spec", "conformance", "fullsize-acceptance.json"), String))
    gate = fixture.scheduled_checkpoints
    scenario = load_scenario(joinpath(root, String(gate.scenario)))
    Tuple(gate.resolution) == scenario.domain.resolution ||
        error("scheduled fixture resolution disagrees with its scenario")
    solver = create_solver(String(gate.solver), scalar_type(scenario))
    initialize!(solver, scenario, NacaFoil(scenario.foil), scenario.seed)
    checkpoint_steps = Dict(round(Int, Float64(value) / scenario.output_dt) => Float64(value) for value in gate.times)
    for step in 1:maximum(keys(checkpoint_steps))
        time = step * scenario.output_dt
        advance!(solver, control_at(scenario, time), scenario.output_dt)
        if haskey(checkpoint_steps, step)
            all(isfinite, export_state(solver).velocity) ||
                error("non-finite scheduled state at t=$time")
            println("passed scheduled checkpoint t=$time")
        end
    end
    return nothing
end

main()
